test/java/seedu/addressbook/commands/AddCommandTest.java

```diff
@@ -0,0 +1,156 @@
+                    tags);
+        } catch (IllegalValueException e) {
+            return;
+        }
+        String error = String.format(
+                "An add command was successfully constructed with invalid input: %s %s %s %s %s %s %s %s",
+                name, phone, isPhonePrivate, email, isEmailPrivate, address, isAddressPrivate, tags);
+        fail(error);
+    }
+
+    @Test
+    public void addCommand_validData_correctlyConstructed() throws Exception {
+        AddCommand cmd = new AddCommand(Name.EXAMPLE, Phone.EXAMPLE, true, Email.EXAMPLE, false,
+                Address.EXAMPLE, true, NO_TAGS); // should never throw
+        ReadOnlyPerson p = cmd.getPerson();
+
+        assertEquals(Name.EXAMPLE, p.getName().fullName);
+        assertEquals(Phone.EXAMPLE, p.getPhone().value);
+        assertTrue(p.getPhone().isPrivate());
+        assertEquals(Email.EXAMPLE, p.getEmail().value);
+        assertFalse(p.getEmail().isPrivate());
+        assertEquals(Address.EXAMPLE, p.getAddress().value);
+        assertTrue(p.getAddress().isPrivate());
+        boolean isTagListEmpty = !p.getTags().iterator().hasNext();
+        assertTrue(isTagListEmpty);
+    }
+
+    @Test
+    public void addCommand_emptyAddressBook_addressBookContainsPerson() {
+        Person p = generateTestPerson();
+        AddCommand cmd = new AddCommand(p);
+        AddressBook book = new AddressBook();
+        cmd.setData(book, LAST_SHOWN_LIST);
+        CommandResult res = cmd.execute();
+        UniquePersonList people = book.getAllPersons();
+
+        assertTrue(people.contains(p));
+        assertEquals(1, people.immutableListView().size());
+        assertFalse(res.getRelevantPersons().isPresent());
+        assertEquals(String.format(AddCommand.MESSAGE_SUCCESS, p), res.feedbackToUser);
+    }
+
+    @Test
+    public void addCommand_addressBookAlreadyContainsPerson_addressBookUnmodified()
+            throws Exception {
+        Person p = generateTestPerson();
+        AddressBook book = new AddressBook();
+        book.addPerson(p); // should never throw
+        AddCommand cmd = new AddCommand(p);
+        cmd.setData(book, LAST_SHOWN_LIST);
+        CommandResult res = cmd.execute();
+
+        assertFalse(res.getRelevantPersons().isPresent());
+        assertEquals(AddCommand.MESSAGE_DUPLICATE_PERSON, res.feedbackToUser);
+        UniquePersonList people = book.getAllPersons();
+        assertTrue(people.contains(p));
+        assertEquals(1, people.immutableListView().size());
+    }
+
+    private static Person generateTestPerson() {
+        try {
+            return new Person(new Name(Name.EXAMPLE), new Phone(Phone.EXAMPLE, false),
+                    new Email(Email.EXAMPLE, true), new Address(Address.EXAMPLE, false), new UniqueTagList());
+        } catch (IllegalValueException e) {
+            fail("test person data should be valid by definition");
+            return null;
+        }
+    }
```

Comment:
Perhaps we can move this to a separate test util class so that it can be shared by other test classes? Test classes often contain duplicate utility methods which is something we should try to avoid.